File: backend/app/services/serpapi/normalizer.py
import datetime
from typing import List, Dict, Any
from urllib.parse import urlparse

def extract_domain(url: str) -> str:
    try:
        parsed = urlparse(url)
        return parsed.netloc.replace("www.", "")
    except Exception:
        return "external-source"

def categorize_source_type(domain: str, url: str) -> str:
    d = domain.lower()
    if d.endswith(".gov") or d.endswith(".gov.in") or d.endswith(".gov.uk"):
        return "government"
    if any(m in d for m in ["reuters.com", "bloomberg.com", "bbc.com", "thehindu.com", "timesofindia", "techcrunch.com"]):
        return "news"
    if any(m in d for m in ["wikipedia.org", "reddit.com", "quora.com", "github.com"]):
        return "community"
    if any(m in d for m in ["linkedin.com", "twitter.com", "x.com", "facebook.com"]):
        return "social_media"
    if "docs." in d or "/docs/" in url:
        return "documentation"
    return "company"

def normalize_serp_results(raw_data: Dict[str, Any], engine: str, query: str) -> List[Dict[str, Any]]:
    items = []
    now = datetime.datetime.utcnow().isoformat()

    if engine == "google":
        organic = raw_data.get("organic_results", [])
        for idx, res in enumerate(organic):
            link = res.get("link")
            if not link:
                continue
            domain = extract_domain(link)
            items.append({
                "id": f"g-{idx}-{now}",
                "title": res.get("title", "Untitled Web Result"),
                "url": link,
                "domain": domain,
                "snippet": res.get("snippet", ""),
                "source_type": categorize_source_type(domain, link),
                "published_at": res.get("date") or res.get("published_date"),
                "search_engine": "google",
                "created_at": now
            })

    elif engine == "google_news":
        news = raw_data.get("news_results", []) or raw_data.get("organic_results", [])
        for idx, res in enumerate(news):
            link = res.get("link") or res.get("share_link")
            if not link:
                continue
            domain = extract_domain(link)
            items.append({
                "id": f"gn-{idx}-{now}",
                "title": res.get("title", "Untitled News Article"),
                "url": link,
                "domain": domain,
                "snippet": res.get("snippet", ""),
                "source_type": "news",
                "published_at": res.get("date") or res.get("published_date"),
                "search_engine": "google_news",
                "created_at": now
            })

    elif engine == "google_maps":
        places = raw_data.get("local_results", []) or raw_data.get("place_results", [])
        for idx, res in enumerate(places):
            link = res.get("website") or (f"https://maps.google.com/?cid={res.get('place_id')}" if res.get("place_id") else "https://maps.google.com")
            domain = extract_domain(link)
            items.append({
                "id": f"gm-{idx}-{now}",
                "title": res.get("title") or res.get("name", "Local Business Listing"),
                "url": link,
                "domain": domain,
                "snippet": f"Location: {res.get('address', 'N/A')} | Rating: {res.get('rating', 'N/A')} stars ({res.get('reviews', 0)} reviews)",
                "source_type": "business_listing",
                "published_at": None,
                "search_engine": "google_maps",
                "created_at": now
            })

    elif engine == "google_jobs":
        jobs = raw_data.get("jobs_results", [])
        for idx, res in enumerate(jobs):
            link = res.get("share_link") or "https://google.com/jobs"
            domain = extract_domain(link)
            items.append({
                "id": f"gj-{idx}-{now}",
                "title": f"{res.get('title')} at {res.get('company_name')}",
                "url": link,
                "domain": domain,
                "snippet": f"Company: {res.get('company_name')} | Location: {res.get('location')} | Posted: {res.get('detected_extensions', {}).get('posted_at', 'Recently')}",
                "source_type": "company",
                "published_at": res.get("detected_extensions", {}).get("posted_at"),
                "search_engine": "google_jobs",
                "created_at": now
            })

    elif engine == "google_shopping":
        shopping = raw_data.get("shopping_results", [])
        for idx, res in enumerate(shopping):
            link = res.get("link") or res.get("product_link")
            if not link:
                continue
            domain = extract_domain(link)
            items.append({
                "id": f"gs-{idx}-{now}",
                "title": res.get("title", "Product Listing"),
                "url": link,
                "domain": domain,
                "snippet": f"Merchant: {res.get('source', domain)} | Price: {res.get('price', 'N/A')} | Rating: {res.get('rating', 'N/A')}",
                "source_type": "business_listing",
                "published_at": None,
                "search_engine": "google_shopping",
                "created_at": now
            })

    return items

File: src/lib/serpapi/normalizer.ts
import { SearchEngineType, SearchResultItem, SourceType } from '@/types';

export function normalizeSearchResult(
  rawData: any,
  engine: SearchEngineType,
  query: string
): SearchResultItem[] {
  const items: SearchResultItem[] = [];
  const now = new Date().toISOString();

  if (engine === 'google') {
    const organic = rawData?.organic_results || [];
    organic.forEach((res: any, idx: number) => {
      if (!res.link) return;
      const domain = extractDomain(res.link);
      items.push({
        id: `google-${idx}-${Date.now()}`,
        title: res.title || 'Untitled Web Result',
        url: res.link,
        domain: domain,
        snippet: res.snippet || res.snippet_highlighted_words?.join(' ') || '',
        source_type: categorizeSourceType(domain, res.link),
        published_at: res.date || res.published_date || null,
        search_engine: 'google',
        raw_data: { position: res.position, rich_snippet: res.rich_snippet },
        created_at: now
      });
    });
  } else if (engine === 'google_news') {
    const newsResults = rawData?.news_results || rawData?.organic_results || [];
    newsResults.forEach((res: any, idx: number) => {
      const link = res.link || res.share_link;
      if (!link) return;
      const domain = extractDomain(link);
      items.push({
        id: `news-${idx}-${Date.now()}`,
        title: res.title || 'Untitled News Article',
        url: link,
        domain: domain,
        snippet: res.snippet || res.summary || '',
        source_type: 'news',
        published_at: res.date || res.published_date || null,
        search_engine: 'google_news',
        raw_data: { source: res.source?.name || res.source },
        created_at: now
      });
    });
  } else if (engine === 'google_maps') {
    const localResults = rawData?.local_results || rawData?.place_results || [];
    localResults.forEach((res: any, idx: number) => {
      const link = res.website || res.links?.website || res.place_id ? `https://maps.google.com/?cid=${res.place_id}` : 'https://maps.google.com';
      const domain = extractDomain(link);
      items.push({
        id: `maps-${idx}-${Date.now()}`,
        title: res.title || res.name || 'Local Business Listing',
        url: link,
        domain: domain,
        snippet: `Location: ${res.address || 'Address unavailable'} | Rating: ${res.rating || 'N/A'} stars (${res.reviews || 0} reviews) | Type: ${res.type || 'Business'}`,
        source_type: 'business_listing',
        published_at: null,
        search_engine: 'google_maps',
        raw_data: { address: res.address, rating: res.rating, reviews: res.reviews, phone: res.phone },
        created_at: now
      });
    });
  } else if (engine === 'google_jobs') {
    const jobsResults = rawData?.jobs_results || [];
    jobsResults.forEach((res: any, idx: number) => {
      const link = res.share_link || res.apply_options?.[0]?.link || res.job_id ? `https://www.google.com/search?q=${encodeURIComponent(query)}#htivrt=jobs` : 'https://google.com/jobs';
      const domain = extractDomain(link);
      items.push({
        id: `jobs-${idx}-${Date.now()}`,
        title: `${res.title} at ${res.company_name}`,
        url: link,
        domain: domain,
        snippet: `Company: ${res.company_name} | Location: ${res.location} | Posted: ${res.detected_extensions?.posted_at || 'Recently'} | Description: ${(res.description || '').slice(0, 200)}...`,
        source_type: 'company',
        published_at: res.detected_extensions?.posted_at || null,
        search_engine: 'google_jobs',
        raw_data: { company: res.company_name, location: res.location, via: res.via },
        created_at: now
      });
    });
  } else if (engine === 'google_shopping') {
    const shoppingResults = rawData?.shopping_results || [];
    shoppingResults.forEach((res: any, idx: number) => {
      const link = res.link || res.product_link || '';
      if (!link) return;
      const domain = extractDomain(link);
      items.push({
        id: `shopping-${idx}-${Date.now()}`,
        title: res.title || 'Product Listing',
        url: link,
        domain: domain,
        snippet: `Merchant: ${res.source || domain} | Price: ${res.price || 'N/A'} | Rating: ${res.rating || 'N/A'} (${res.reviews || 0} reviews)`,
        source_type: 'business_listing',
        published_at: null,
        search_engine: 'google_shopping',
        raw_data: { price: res.price, merchant: res.source, rating: res.rating },
        created_at: now
      });
    });
  }

  return deduplicateResults(items);
}

export function extractDomain(url: string): string {
  try {
    const parsed = new URL(url);
    return parsed.hostname.replace(/^www\./, '');
  } catch {
    return 'external-source';
  }
}

export function categorizeSourceType(domain: string, url: string): SourceType {
  const d = domain.toLowerCase();
  if (d.endsWith('.gov') || d.endsWith('.gov.in') || d.endsWith('.gov.uk') || d.includes('official.gov')) {
    return 'government';
  }
  if (
    d.includes('reuters.com') ||
    d.includes('bloomberg.com') ||
    d.includes('bbc.com') ||
    d.includes('thehindu.com') ||
    d.includes('timesofindia') ||
    d.includes('techcrunch.com') ||
    d.includes('moneycontrol.com') ||
    d.includes('economic-times')
  ) {
    return 'news';
  }
  if (
    d.includes('wikipedia.org') ||
    d.includes('reddit.com') ||
    d.includes('quora.com') ||
    d.includes('github.com') ||
    d.includes('medium.com')
  ) {
    return 'community';
  }
  if (d.includes('linkedin.com') || d.includes('twitter.com') || d.includes('x.com') || d.includes('facebook.com')) {
    return 'social_media';
  }
  if (d.includes('docs.') || d.includes('developer.') || url.includes('/docs/')) {
    return 'documentation';
  }
  return 'company';
}

export function deduplicateResults(results: SearchResultItem[]): SearchResultItem[] {
  const seenUrls = new Set<string>();
  const seenTitles = new Set<string>();
  const unique: SearchResultItem[] = [];

  for (const item of results) {
    const normalizedUrl = item.url.toLowerCase().split('#')[0].replace(/\/$/, '');
    const normalizedTitle = item.title.toLowerCase().trim().replace(/[^a-z0-9]/g, '');

    if (seenUrls.has(normalizedUrl) || seenTitles.has(normalizedTitle)) {
      continue;
    }

    seenUrls.add(normalizedUrl);
    seenTitles.add(normalizedTitle);
    unique.push(item);
  }

  return unique;
}
